datalad/support/gitrepo.py

```diff
@@ -2355,17 +2358,67 @@ def get_submodules_(self, paths=None):
             return
 
         modinfo = self._parse_gitmodules()
-        for path, props in self.get_content_info(
-                paths=paths,
-                ref=None,
-                untracked='no').items():
-            if props.get('type', None) != 'dataset':
+        if not modinfo:
+            # we exit early, if there is nothing on record (even though
+            # a .gitmodules file exists).
+            # without this conditional exit, we would be able to discover
+            # subprojects even when they are not recorded in .gitmodules,
+            # but at the cost of running an unconstrained ls-files call
+            # below
+            return
+
+        # taken from 3.9's pathlib, can be removed once the minimally
+        # supported python version
+        def is_relative_to(self, *other):
```

Comment:
note: I was curious if we have other uses, grepped to find added in 51e061135a

```
datalad/support/gitrepo.py-                # this could potentially be expensive if lots of files become
datalad/support/gitrepo.py-                # directories, but it is unlikely to happen often
datalad/support/gitrepo.py:                # Note: PurePath.is_relative_to was added in 3.9 and seems slowish
datalad/support/gitrepo.py-                # path_is_subpath faster, also if comparing to "in f.parents"
datalad/support/gitrepo.py-                f_str = str(f)
datalad/support/gitrepo.py-                if any(path_is_subpath(str(f2), f_str) for f2 in status_state['added']):
```

so if speed is of concern, might be worth timing and extracting into a reusable helper.

Reply:
I am not interested in improving upon this function. I was planing on removing it whenever we reach PY3.9

Reply:
> path_is_subpath() is practically unused outside the tests,

```shell
$> git grep -p 'path_is_subpath(' | grep -v test_
datalad/interface/results.py=def results_from_annex_noinfo(ds, requested_paths, respath_by_status, dir_fail_msg,
datalad/interface/results.py:                if path_is_subpath(fp, p)]
datalad/interface/results.py:                    if path_is_subpath(fp, p)]
datalad/interface/utils.py=def discover_dataset_trace_to_targets(basepath, targetpaths, current_trace,
datalad/interface/utils.py:                       path_is_subpath(t, current_trace[-1]) and
datalad/metadata/aggregate.py=def _get_dsinfo_from_aggmetadata(ds_path, path, recursive, db):
datalad/metadata/aggregate.py:        if path_is_subpath(agginfo_path, seed_ds):
datalad/metadata/aggregate.py:            elif any(path_is_subpath(ep, rp) for ep in exclude):
datalad/metadata/metadata.py=def _get_containingds_from_agginfo(info, rpath):
datalad/metadata/metadata.py:             if path_is_subpath(rpath, subds)],
datalad/metadata/metadata.py=def legacy_query_aggregated_metadata(reporton, ds, aps, recursive=False,
datalad/metadata/metadata.py:                              path_is_subpath(sub, rpath)]
datalad/metadata/metadata.py=class Metadata(Interface):
datalad/metadata/metadata.py:                if parentds and not path_is_subpath(dspath, parentds[-1]):
datalad/support/gitrepo.py=class GitRepo(CoreGitRepo):
datalad/support/gitrepo.py:                if any(path_is_subpath(str(f2), f_str) for f2 in status_state['added']):
datalad/support/parallel.py=def no_parentds_in_futures(futures, path, skip=tuple()):
datalad/support/parallel.py:    return all(not path_is_subpath(path, p) or p in skip for p in futures)
datalad/support/parallel.py=def no_subds_in_futures(futures, path, skip=tuple()):
datalad/support/parallel.py:    return all(not path_is_subpath(p, path) or p in skip for p in futures)
datalad/utils.py=def path_startswith(path, prefix):
datalad/utils.py:def path_is_subpath(path, prefix):
datalad/utils.py=def import_module_from_file(modpath, pkg=None, log=lgr.debug):
datalad/utils.py:            if path_is_subpath(modpath, pkgpath):
```

and I think it is just a matter of fixing the tests, so submitted https://github.com/datalad/datalad/pull/6955 . Let's see how it behaves on windows